### PR TITLE
Fix for issue #7413 - Allow non-zero exit codes for inspec provisioner

### LINKED ISSUE #7413: Allow non-zero exit codes for inspec provisioner
Inspec has a non-zero code for tests with skipped controls that do not match or apply to the current system. Here https://github.com/inspec/inspec/issues/1825#issuecomment-382650911
```
0 run ok + all passed:
100 run okay + failures`
101 run okay + skipped only`
```
#7385 adds exit codes for shell and windows provisioners and I was hoping the same could be done for inspec.

For example running the [linux patch baseline profile](https://github.com/dev-sec/linux-patch-baseline) on a fully patched machine will return a 101 exit code because no patches need to be applied
``` amazon-ebs: [2019-03-15T13:10:36-04:00] DEBUG: Registering rule verify-patches
    amazon-ebs: [2019-03-15T13:10:36-04:00] DEBUG: Registering rule patches
    amazon-ebs: [2019-03-15T13:10:36-04:00] DEBUG: Registering rule os-patches
    amazon-ebs:
    amazon-ebs: Profile: DevSec Linux Patch Benchmark (linux-patch-baseline)
    amazon-ebs: Version: 0.4.0
    amazon-ebs: Target:  ssh://root@127.0.0.1:33148
    amazon-ebs:
    amazon-ebs:   ✔  verify-patches: Operating system is up-to-date
    amazon-ebs:      ✔  Linux Update should be uptodate
    amazon-ebs:   ↺  patches: All operating system package updates are installed
    amazon-ebs:      ↺  Skipped control due to only_if condition.
    amazon-ebs:
    amazon-ebs:
    amazon-ebs: Profile Summary: 1 successful control, 0 control failures, 1 control skipped
    amazon-ebs: Test Summary: 1 successful, 0 failures, 1 skipped
```
and the resulting error
```
2019/03/15 13:11:07 [INFO] (telemetry) ending amazon-ebs
2019/03/15 13:11:07 ui error: Build 'amazon-ebs' errored: Error executing Inspec: Non-zero exit status: exit status 101
2019/03/15 13:11:07 Builds completed. Waiting on interrupt barrier...
2019/03/15 13:11:07 machine readable: error-count []string{"1"}
2019/03/15 13:11:07 ui error: 
==> Some builds didn't complete successfully and had errors:
2019/03/15 13:11:07 machine readable: amazon-ebs,error []string{"Error executing Inspec: Non-zero exit status: exit status 101"}
2019/03/15 13:11:07 ui error: --> amazon-ebs: Error executing Inspec: Non-zero exit status: exit status 101
==> Builds finished but no artifacts were created.
2019/03/15 13:11:07 [INFO] (telemetry) Finalizing.
Build 'amazon-ebs' errored: Error executing Inspec: Non-zero exit status: exit status 101

==> Some builds didn't complete successfully and had errors:
--> amazon-ebs: Error executing Inspec: Non-zero exit status: exit status 101

==> Builds finished but no artifacts were created.
```

### PR DESCRIPTION
**Pull request by finchr**
I ran into the same problem addressed by #7413 and took a stab at fixing it.

I found and applied the fix from the PR https://github.com/jrbeilke/packer-provisioner-inspec/pull/11 to this repo and it appears to address the problem.

I will remove the WIP when I have figured out how to write a test case.

Closes #7413

### CODE REVIEW

**Comment by finchr** [comment]
I just pushed a new test for the valid exit codes.

**Comment by rfinch-redviolet** [comment]
All the tests have passed, just waiting on approval. What are the next steps?

**Comment by finchr** [comment]
I fixed a bug I found the getVersion method and pushed a fix for it and the getVersion test. I also fixed the PACKE_ACC=1 test cases so they all pass now.

File: provisioner/inspec/provisioner.go [hunk 1]
```diff
@@ -412,9 +418,33 @@ func (p *Provisioner) executeInspec(ui packersdk.Ui, comm packersdk.Communicator
...
+				// Check exit code against allowed codes (likely just 0)
+				validExitCode := false
+				for _, v := range p.config.ValidExitCodes {
+					if exitStatus == v {
+						validExitCode = true
+					}
+				}
+				if !validExitCode {
+					return fmt.Errorf(
+						"Inspec exited with non-zero exit status: %d. Allowed exit codes are: %v",
```
Comment (azr) [inline provisioner/inspec/provisioner.go:441]:
```suggestion
						"Inspec exited with unexpected exit status: %d. Allowed exit codes are: %v",
```

Comment (mbainter) [reply provisioner/inspec/provisioner.go:441]: Perhaps replace "Allowed exit codes" with "expected exit codes" as well, so it doesn't imply that only those codes are allowed.

Comment (rfinch-redviolet) [reply provisioner/inspec/provisioner.go]: > Perhaps replace "Allowed exit codes" with "expected exit codes" as well, so it doesn't imply that only those codes are allowed.

Done, I just pushed the change.

Comment (rfinch-redviolet) [reply provisioner/inspec/provisioner.go]: @azr I added your suggestion as well.

**Review by sylviamoss — COMMENTED** [review]
This looks good! The only thing missing is the documentation. 
You should add the documentation for the new configuration in https://github.com/hashicorp/packer/blob/master/website/content/docs/provisioners/inspec.mdx

**Comment by finchr** [comment]
> This looks good! The only thing missing is the documentation.
> You should add the documentation for the new configuration in https://github.com/hashicorp/packer/blob/master/website/content/docs/provisioners/inspec.mdx

Thanks for the link @sylviamoss, I didn't know where the documentation lived in the repo! I'll have an update in sometime today.

**Comment by finchr** [comment]
Hi @sylviamoss I just pushed a change with documentation. I also changed the default values for `valid_exit_codes` to avoid errors with skipped inspec tests.

File: provisioner/inspec/provisioner.go [hunk 2]
```diff
@@ -157,6 +159,10 @@ func (p *Provisioner) Prepare(raws ...interface{}) error {
 		errs = packersdk.MultiErrorAppend(errs, fmt.Errorf("user: could not determine current user from environment."))
 	}
 
+	if p.config.ValidExitCodes == nil {
+		p.config.ValidExitCodes = []int{0, 101}
```
Comment (sylviamoss) [inline provisioner/inspec/provisioner.go:163]: https://docs.chef.io/inspec/cli/#exec leaving here some docs of why 101. 
`101  normal exit, at least one test skipped but none failed`

**Review by sylviamoss — APPROVED** [review]
Nice! Thanks for updating the docs. This LGTM!

**Pull request merged**